engine/common/mod_bmodel.c

```diff
@@ -207,6 +210,151 @@ static mlumpinfo_t		extlumps[EXTRA_LUMPS] =
+	size_t foundPathBufferLength )
+{
+	char pathInWad[MAX_VA_STRING];
+	int wadIndex;
+
+	if( !list || !textureName || !(*textureName) )
+	{
+		return -1;
+	}
+
+	// check wads in reverse order
+	for( wadIndex = list->count - 1; wadIndex >= 0; --wadIndex )
+	{
+		Q_snprintf( pathInWad, sizeof(pathInWad), "%s.wad/%s.mip", list->wadnames[wadIndex], textureName );
+
+		if( FS_FileExists( pathInWad, false ) )
+		{
+			if( foundPathBuffer && foundPathBufferLength > 0 )
+			{
+				Q_strncpy( foundPathBuffer, pathInWad, foundPathBufferLength );
+			}
+
+			return wadIndex;
+		}
+	}
+
+	return -1;
+}
+
+static size_t Mod_CalculateMipTexSize( mip_t* mipTex, qboolean customPalette )
+{
+	if( !mipTex )
+	{
+		return 0;
+	}
+
+	return sizeof(*mipTex) +
+		(( mipTex->width * mipTex->height * 85) >> 6) +
+		(customPalette ? MIPTEX_CUSTOM_PALETTE_SIZE_BYTES : 0);
+}
+
+static qboolean Mod_CalcMipTexUsesCustomPalette( dbspmodel_t* bmod, int textureIndex )
+{
+	int size = 0;
+	int nextTextureIndex = 0;
+	mip_t* mipTex = NULL;
+
+	mipTex = Mod_GetMipTexForTexture( bmod, textureIndex );
+
+	if( !mipTex || mipTex->offsets[0] <= 0 )
+	{
+		return false;
+	}
+
+	// Calculate the size assuming we are not using a custom palette.
+	size = (int)Mod_CalculateMipTexSize( mipTex, false );
+
+	// Compute next data offset to determine allocated miptex space
+	for( nextTextureIndex = textureIndex + 1; nextTextureIndex < loadmodel->numtextures; ++nextTextureIndex )
+	{
+		int nextOffset = bmod->textures->dataofs[nextTextureIndex];
+
+		if( nextOffset != -1 )
+		{
+			int remainingBytes = nextOffset - (bmod->textures->dataofs[textureIndex] + size);
+			return remainingBytes >= MIPTEX_CUSTOM_PALETTE_SIZE_BYTES;
+		}
+	}
+
+	// There was no other miptex after this one.
+	// See if there is enough space between the end and our offset.
+	return bmod->texdatasize - (bmod->textures->dataofs[textureIndex] + size) >= MIPTEX_CUSTOM_PALETTE_SIZE_BYTES;
+}
+
+static qboolean Mod_NameImpliesTextureIsAnimated( texture_t* tex )
+{
+	if( !tex )
+	{
+		return false;
+	}
+
+	if( tex->name[0] != '-' && tex->name[0] != '+' )
+	{
+		// Not an animated texture name
+		return false;
+	}
+
+	// Name implies texture is animated - check second character is valid.
+
+	if( !(tex->name[1] >= '0' && tex->name[1] <= '9') &&
+	    !(tex->name[1] >= 'a' && tex->name[1] <= 'j') )
+	{
+		Con_Printf( S_ERROR "Mod_NameImpliesTextureIsAnimated: animating texture \"%s\" has invalid name\n", tex->name );
+		return false;
+	}
+
+	return true;
+}
+
+static void Mod_CreateDefaultTexture(texture_t** texture)
+{
+	// Pointer must be valid, and value pointed to must be null.
+	if( !texture || *texture != NULL )
+	{
+		return;
+	}
+
+	*texture = Mem_Calloc( loadmodel->mempool, sizeof(texture_t) );
+	Q_strncpy( (*texture)->name, REF_DEFAULT_TEXTURE, sizeof((*texture)->name) );
+
+#if !XASH_DEDICATED
+	if( !Host_IsDedicated() )
+	{
+		(*texture)->gl_texturenum = R_GetBuiltinTexture( REF_DEFAULT_TEXTURE );
```

Comment:
To avoid this unreadable mess, let's make a variable `texture_t *tex`, allocate a texture, initialize all fields and then set `*texture = tex` at once.